0019-refactor-image-operations/README.md

```diff
@@ -0,0 +1,315 @@
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss that.
+-->
+
+oras-go does not provide a blob storage natively, however the ORAS CLI does. While it is marked as internal, it is simple to vendor into the Zarf project. Additionally, issue [#881](https://github.com/oras-project/oras-go/issues/881) in oras-go requests caching as part of the library. The maintainers have noted that it seems like a valuable feature to add.
+
+oras-go does not natively support pulling images from the Docker daemon. Zarf will instead pull from the Docker daemon directly, which results in an OCI formatted tar file. Once extracted into a directory it can be treated as a normal oci-layout for use with the oras-go library. 
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept proposals with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Zarf testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://docs.zarf.dev/contribute/testing/
+-->
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this proposal.
+
+Since image pulling and pushing is a core functionality of Zarf, most of the end to end tests will, by nature, test image pulling and pushing. Still, there should be a test to ensure a package built with a prior version of Zarf is able to push of all it's images successfully. 
+
+Additionally, the image push should have unit tests, and the image pull unit tests should be improved. 
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until you're targeting a release.*
+
+Define what needs to happen for this feature to move from alpha to beta to GA
+(General Availability). Focus on key signals or criteria that show the feature
+is ready for each stage.
+
+Consider the following stages when setting graduation criteria:
+- Alpha: Feature is behind a feature flag, basic tests in place.
+- Beta: Gather feedback from users, complete core features, add more tests.
+- GA: Prove real-world usage, complete rigorous testing, gather feedback.
+
+In general, features should wait at least two releases between Beta and GA to
+allow time for feedback. For features moving to GA, include conformance tests
+to ensure stability and compatibility.
+
+#### Deprecation
+If this feature will eventually be deprecated, plan for it:
+- Announce deprecation and support policy.
+- Wait at least two versions before fully removing it.
+-->
+
+As this is not a feature, but a refactor of an implementation this change will be introduced as stable immediately. 
```

Comment:
I think that we should strongly consider feature flagging this during implementation, so changes can be A/B tested. We can always opt to remove it in favor of a full unflagged migration when making it available to users. Just want to make sure we have as much flexibility to explore changes that improve the architecture and behavior.

Reply:
I do think it would be possible, but IMO the added complexity isn't worth it, at least yet. It gets tricky because in order for a real gain to be seen by doing this, we would have to start concurrently pulling images and image layers. Even though ORAS does handle pulling multiple images, it takes some work on our part and makes the code considerably more complex.

I will mention this in the alternatives.

Reply:
That's a reasonable idea. Right now we have the following situations:

- package create with older version of Zarf (using Crane) must deploy with newer version of Zarf (using ORAS)
- package create with a newer version of Zarf (using ORAS) must deploy with an older version of Zarf (using Crane). 

I'd like to avoid introducing additional flags / options for users, but feature flagging could help for testing packages built with prior versions of Zarf. Another option would be downloading a previous version of Zarf in our e2e suite. This won't be the last time we have a change like this, so I'll mark it as a topic worth synchronous discussion with the team on what the optimal testing here looks like.